engine/instant_query.go

```diff
@@ -15,32 +17,89 @@ import (
 
 type instantQuery struct {
 	plan model.VectorOperator
+	pool *model.VectorPool
+	expr parser.Expr
+	ts   time.Time
 }
 
-func newInstantQuery(plan model.VectorOperator) promql.Query {
-	return &instantQuery{plan: plan}
+func newInstantQuery(plan model.VectorOperator, pool *model.VectorPool, expr parser.Expr, ts time.Time) promql.Query {
+	return &instantQuery{
+		plan: plan,
+		pool: pool,
+		expr: expr,
+		ts:   ts,
+	}
 }
 
 func (q *instantQuery) Exec(ctx context.Context) *promql.Result {
-	return nil
+	// Handle case with strings early on as this does not need us to process samples.
+	// TODO(saswatamcode): Modify models.StepVector to support all types and check during plan creation.
+	switch e := q.expr.(type) {
+	case *parser.StringLiteral:
+		return &promql.Result{Value: promql.String{V: e.Val, T: q.ts.UnixMilli()}}
+	}
+
+	vs, err := q.plan.Next(ctx)
+	if err != nil {
+		return newErrResult(err)
+	}
+
+	if len(vs) == 0 {
+		return &promql.Result{}
+	}
+
+	resultSeries, err := q.plan.Series(ctx)
+	if err != nil {
+		return newErrResult(err)
+	}
+
+	series := make([]promql.Series, len(resultSeries))
+	for i := 0; i < len(resultSeries); i++ {
+		series[i].Metric = resultSeries[i]
+	}
+
+	for _, vector := range vs {
+		for _, sample := range vector.Samples {
+			if len(series[sample.ID].Points) == 0 {
+				series[sample.ID].Points = make([]promql.Point, 0, 1)
```

Comment:
Sorry one more change here, can we move this bellow line 58 and avoid the `if` condition altogether?